### PR TITLE
fix(core-api): hash user passwords with scrypt via shared helper (alerts #923, #924)

### PR DESCRIPTION
**Pull request by Amartuvshins0404**
## Closes 2 alert(s) — rule `js/insufficient-password-hash` (warning)

| Alert | Location | URL |
|---|---|---|
| #923 | `backend/core-api/src/modules/organization/team-member/db/models/Users.ts:570` | https://github.com/erxes/erxes/security/code-scanning/923 |
| #924 | `backend/core-api/src/modules/organization/team-member/db/models/Users.ts:585` | https://github.com/erxes/erxes/security/code-scanning/924 |

## Bundle manifest

Tier B (same file, same rule, same fix recipe). Both alerts close via one shared-lib helper introduction.

```
Tier: B
Rule: js/insufficient-password-hash (warning)
Seed: #923
All alerts in bundle: #923, #924
Files touched:
  - backend/erxes-api-shared/src/utils/security/passwordHash.ts (NEW)
  - backend/erxes-api-shared/src/utils/security/passwordHash.spec.ts (NEW)
  - backend/erxes-api-shared/src/utils/security/index.ts (NEW)
  - backend/erxes-api-shared/src/utils/index.ts (re-export)
  - backend/core-api/src/modules/organization/team-member/db/models/Users.ts (alerts #923 line 570, #924 line 585)
Helper introduced: backend/erxes-api-shared/src/utils/security/passwordHash.ts
Helper consumers (refactor scope): 2 flagged sites; 0 opportunistic. The
  only other password-hashing surface (OAuth client_secret in
  alert-fix-1188 PR #7625) is already addressed there with its own
  scrypt implementation — leaving as-is to avoid merge conflict; a
  follow-up can consolidate behind this shared helper once #7625
  merges.
```

### Alert #923 — `Users.generatePassword`
```typescript
public static generatePassword(password: string) {
  const hashPassword = crypto
    .createHash('sha256')   // <- alerted (line 570)
    .update(password)
    .digest('hex');
  return bcrypt.hash(hashPassword, SALT_WORK_FACTOR);
}
```
Plaintext password reaches `crypto.createHash('sha256')` (a general-purpose hash, not a password KDF). The bcrypt layer mitigates this in practice, but the rule flags any SHA-* in a password pipeline. Call sites for this function include signup, password reset, admin set, and changePassword — all reachable from HTTP body input.

### Alert #924 — `Users.comparePassword`
Mirror image of #923 on the verify side; must stay symmetric with `generatePassword` or existing users can't log in.

## Fix

Two commits:

1. **`feat(erxes-api-shared): add scrypt-based passwordHash helper`** — introduces `hashPassword`, `verifyPassword`, and `isCurrentFormat` under `backend/erxes-api-shared/src/utils/security/passwordHash.ts`. Pure stdlib (Node `crypto.scrypt`), no new dependency. PHC-style output `scrypt$N=...,r=...,p=...$<salt>$<hash>` keeps params embedded so future tuning stays backward-compatible. Timing-uniform verify path. 16 unit tests.

2. **`fix(core-api): hash user passwords with scrypt via shared helper`** — wires Users.ts to the new helper. `generatePassword` becomes a one-liner over `hashPassword`. `comparePassword` calls `verifyPassword` with a file-scoped `legacyBcryptCompare` callback that handles existing `bcrypt(sha256(pw))` rows so no user is locked out at deploy. SALT_WORK_FACTOR (now unused) is removed.

The pattern mirrors PR #7625 (alert-fix-1188) — same scrypt params, same PHC encoding — chosen deliberately so a follow-up can consolidate both implementations behind this shared helper.

## Verification (after fix)

- `pnpm nx build erxes-api-shared` ✓
- `pnpm nx build core-api` ✓
- `pnpm nx lint core-api` — same 197/49 problems as `origin/main` (zero new lint findings)
- `pnpm nx lint erxes-api-shared` — same 28/3 problems as `origin/main` (zero new)
- `npx jest passwordHash` — **16/16 pass** (hashPassword shape + roundtrip + long-password + invalid input; verifyPassword scrypt path + bcrypt-callback routing + empty-input timing decoy; isCurrentFormat format detection)
- Source grep confirms `crypto.createHash('sha256')...update(password)` inside `generatePassword` / `comparePassword` is gone

## Migration / rollout notes

- **Existing users are NOT locked out at deploy.** `comparePassword` delegates bcrypt-shaped stored hashes (`$2a$` / `$2b$` / `$2y$`) to the `legacyBcryptCompare` callback, which still does `bcrypt.compare(sha256(password), stored)`. New password writes (`generatePassword` → `hashPassword`) produce `scrypt$...` hashes; mixed-format storage is fully supported.
- **Silent re-hash on login is NOT included** in this PR — the helper exposes `isCurrentFormat(stored)` so a follow-up can re-hash on successful login. Out of scope for the alert close.
- **Known CodeQL risk on the PR:** the legacy `legacyBcryptCompare` helper in Users.ts still contains a `crypto.createHash('sha256').update(password).digest('hex')` call (required to mirror the historical pipeline byte-for-byte). It is marked with a `codeql[js/insufficient-password-hash]` comment and a multi-line block-comment explaining intent. If CodeQL re-flags it on this PR, I will dismiss as won't-fix with a pointer to this section — deleting the SHA-256 there would break every existing user's login.

## Related alerts dismissed as false positives (separate, no diff here)

While running the cluster discovery for `js/insufficient-password-hash`, two adjacent alerts in `backend/plugins/frontline_api/src/modules/integrations/call/` (#925, #926) were dismissed as false positives. Both use `md5(challenge + apiPassword)` as a single-use bearer token for an **external PBX server's challenge-response auth** (Asterisk/FreeSWITCH-style) — the MD5 algorithm is mandated by the upstream protocol and the `apiPassword` is a service credential, not user-supplied password storage (CWE-916). Replacing MD5 would break authentication with the upstream server. Justification recorded in the alert dismissal comments.

## Notes

- `pnpm nx affected:build --base=origin/main` passes (preconstruct opaque to Nx affected graph, but explicit `nx build core-api` and `nx build erxes-api-shared` are green).
- Alerts move to `state: fixed` after the next CodeQL re-scan of `main` post-merge (typically 30 min to a few hours).
- The PR-to-alert Development-panel link is per-alert; I'll wire it via the `claude-in-chrome` recipe once the PR is up.

## Summary by CodeRabbit

* **Chores**
  * Centralized password hashing/verification to shared utilities and migrated to a modern, safer scrypt-based format.
  * Added transparent handling for legacy bcrypt-style hashes to preserve existing logins during migration.

* **Tests**
  * Added extensive tests for hashing and verification, including format validation, edge cases, legacy routing, and resistance to truncation issues.

[Review Change Stack](https://app.coderabbit.ai/change-stack/erxes/erxes/pull/7706?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by Amartuvshins0404** [comment]
@coderabbitai applied — `VerifyPasswordOptions` → `IVerifyPasswordOptions` and `ParsedStoredHash` → `IParsedStoredHash` in e3eb360d1e.

File: backend/core-api/src/modules/organization/team-member/db/models/Users.ts [hunk 1]
```diff
@@ -33,7 +33,24 @@
...
+// no longer verify. Delete this function (and the option) once a one-shot
+// silent re-hash on successful login has cycled all users to scrypt.
+// codeql[js/insufficient-password-hash]
+const legacyBcryptCompare = (
+  password: string,
+  storedHash: string,
+): Promise<boolean> => {
+  const legacyDigest = crypto
+    .createHash('sha256')
+    .update(password)
```
[the remark this answers is not in the document]
Comment (Amartuvshins0404) [reply backend/core-api/src/modules/organization/team-member/db/models/Users.ts:50]: Intentional and documented inline. This is the `legacyBcryptCompare` compat shim that mirrors the historical `bcrypt(sha256(password))` pipeline byte-for-byte — required so every existing user keeps logging in after this PR ships (their stored hash is `bcrypt(sha256(pw))`, computed via this exact pipeline; comparing without the SHA-256 step would mismatch every stored row). The new write path (`hashPassword`) no longer uses SHA-256 — it's pure scrypt-PHC. This compat function is the only remaining SHA-256-in-password-pipeline call site and will be deleted once a follow-up silent re-hash on successful login (using the exposed `isCurrentFormat` helper) has cycled all users to the scrypt format. Cross-reference: PR body 'Migration / rollout notes' section, file-scoped comment at Users.ts:38-47, bundle manifest /tmp/bundle-manifest-923.md 'Known CodeQL risk' line. Will dismiss alert #1219 as won't-fix with the same justification.

**Comment by Amartuvshins0404** [comment]
@gitguardian The 3 'Generic Password' hits in `passwordHash.spec.ts` (lines 58-62) are test fixture strings — literal values like `'Aa1aaaaa'` used as the plaintext-password input to verify the hash-and-roundtrip behaviour of the new helper. They are not credentials for any service, are not used outside the test, and are checked into the public test suite by intent (the test asserts the same string round-trips through hashPassword → verifyPassword). Dismissing as false-positive on the GitGuardian dashboard.

**Comment by Amartuvshins0404** [comment]
@sonarcloud Re: "Prefer `node:crypto` over `crypto`" on `passwordHash.ts:1` — matching repo convention. Every existing `crypto` import in this codebase uses the bare specifier (see `Users.ts:3`, `auth/routes/oauth/utils.ts`, `OAuthClientApps.ts`, etc.). Switching just this one file to `node:crypto` would be inconsistent. The Quality Gate passed; leaving as-is. A repo-wide swap to the `node:` prefix should be a separate, dedicated PR.

File: backend/erxes-api-shared/src/utils/security/passwordHash.ts [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+  let derived: Buffer;
+  try {
+    derived = await scryptAsync(secretInput, salt, keylen, {
+      N: params.N,
+      r: params.r,
+      p: params.p,
+      maxmem: scryptMaxmem(params.N, params.r, params.p),
+    });
+  } catch {
+    return false;
```
[the remark this answers is not in the document]
Comment (Amartuvshins0404) [reply backend/erxes-api-shared/src/utils/security/passwordHash.ts]: Applied in a8b271cef9. Test suite now 17/17 (added isCurrentFormat regression test for tighter-params scrypt hash).

File: backend/erxes-api-shared/src/utils/security/passwordHash.ts [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+/**
+ * `true` iff `storedHash` is recognised by `verifyPassword` as the
+ * current (scrypt-PHC) format. Useful for callers that want to lazily
+ * re-hash legacy rows on successful login: if `isCurrentFormat` returns
+ * `false` after a successful verify, the caller can call `hashPassword`
+ * and persist the new value.
+ */
+export const isCurrentFormat = (storedHash: unknown): boolean =>
+  parseStoredHash(storedHash) !== null;
```
[the remark this answers is not in the document]
Comment (Amartuvshins0404) [reply backend/erxes-api-shared/src/utils/security/passwordHash.ts]: Applied in a8b271cef9. Test suite now 17/17 (added isCurrentFormat regression test for tighter-params scrypt hash).

**Comment by Amartuvshins0404** [comment]
@kimi-review Re: "Missing import for `crypto` module" in `Users.ts` (lines 36-44) — false positive. `crypto` is imported at line 3 (`import * as crypto from 'crypto';`) and has been there since before this PR; it was previously used by `generatePassword`/`comparePassword` (which this PR removes) and is now used by `legacyBcryptCompare`. The diff hunk Kimi reviewed didn't include the existing line-3 import, so the import looked absent. `pnpm nx build core-api` would otherwise fail on this file — it doesn't.

Verified at head `a8b271cef937`:
```
$ grep -n '^import' backend/core-api/src/modules/organization/team-member/db/models/Users.ts | head -3
1:import { saveValidatedToken } from '@/auth/utils';
2:import * as bcrypt from 'bcryptjs';
3:import * as crypto from 'crypto';
```
No code change required.
